### PR TITLE
Break Solution-checksums into a checksums for pure-state vs source-generator state

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The goal here is to allow OOP calls (and syncing) and workspaces/solutions/snapshots/compilations/etc. to be more strongly typed.  THe issue today is that we have calls to oop that are basically syaing:

1. run against this solution snapshot *without generators being run* and
2. run against this solution snapshot *with generators being run*.

The former is all the calls taht need to call out to oop to generate SG code in the first place.  The latter is for calls that want to know at least some amount of generated code for (whihc in the future we want to allow to be stale).  

Today, the mechanisms for all of these are the same.  By breaking things up like this, we can more cleanly separate how these components work. ANd we can very much isolate out all the codepaths that want to sync/execute using *no* generator info.  

This does mean an additional hop in communication.  But that should be minor.  Either we're doing a bulk sync, and a single message is nothing. Or we're already doing a tiny handful, and a single message is just noise.

### CODE REVIEW

File: src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs [hunk 1]
```diff
@@ -119,7 +119,7 @@ public ChecksumObjectCollection<ProjectStateChecksums> ToProjectObjects(Checksum
         public ChecksumObjectCollection<DocumentStateChecksums> ToDocumentObjects(ChecksumCollection collection)
             => new(this, collection);
 
-        internal async Task VerifyAssetAsync(SolutionStateChecksums solutionObject)
+        internal async Task VerifyAssetAsync(SolutionStateChecksums1 solutionObject)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs:122]: renamed only so i could audit all locations.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski this is ready for review.

File: src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs [hunk 2]
```diff
@@ -220,8 +220,13 @@ internal async Task VerifyAssetAsync(DocumentStateChecksums documentObject)
 
         internal async Task VerifySolutionStateSerializationAsync(Solution solution, Checksum solutionChecksum)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs:221]: i would recommend looking at the product changes first before test changes.

File: src/Workspaces/Core/Portable/Remote/WellKnownSynchronizationKind.cs [hunk 3]
```diff
@@ -8,7 +8,11 @@ internal enum WellKnownSynchronizationKind
 {
     // Start at a different value from 0 so that if we ever get 0 we know it's a bug.
 
-    SolutionState = 1,
+    // Solution snapshot state, including generator info (like frozen generated documents).
+    SolutionCompilationState = 1,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Remote/WellKnownSynchronizationKind.cs:12]: New 'topmost checksum sync object kind'.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 4]
```diff
@@ -75,7 +75,7 @@ internal sealed partial class SolutionCompilationState
         _frozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument;
 
         // when solution state is changed, we recalculate its checksum
-        _lazyChecksums = AsyncLazy.Create(c => ComputeChecksumsAsync(projectsToInclude: null, c));
+        _lazyChecksums = AsyncLazy.Create(c => ComputeChecksumsAsync(projectId: null, c));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:78]: we don't need projectsToInclude as the data we incorporate into the solution-compilation-state is not affected by the set of projects included, only the single project we ask to sync.

when we get to the *solution state* however, this bubbles into N projectsToInclude.  NAmely the transitive set of proejcts taht the sync'ed project depends on .  

so you'll stills see ComputeChecksumsAsync(projectsToInclude) in SolutionState, just not in SolutionCompilationState.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_Checksum.cs [hunk 5]
```diff
@@ -20,6 +20,116 @@
...
+        /// </summary>
+        private readonly AsyncLazy<SolutionCompilationStateChecksums> _lazyChecksums;
+
+        /// <summary>
+        /// Mapping from project-id to the checksums needed to synchronize it (and the projects it depends on) over 
+        /// to an OOP host.  Lock this specific field before reading/writing to it.
+        /// </summary>
+        private readonly Dictionary<ProjectId, AsyncLazy<SolutionCompilationStateChecksums>> _lazyProjectChecksums = new();
+
+        public bool TryGetStateChecksums([NotNullWhen(true)] out SolutionCompilationStateChecksums? stateChecksums)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_Checksum.cs:37]: similar pattern to SolutionState.XXXChecksum methods.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_Checksum.cs [hunk 6]
```diff
@@ -20,6 +20,116 @@
...
+            ProjectId? projectId,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                using (Logger.LogBlock(FunctionId.SolutionCompilationState_ComputeChecksumsAsync, this.SolutionState.FilePath, cancellationToken))
+                {
+                    var solutionStateChecksum = projectId == null
+                        ? await this.SolutionState.GetChecksumAsync(cancellationToken).ConfigureAwait(false)
+                        : await this.SolutionState.GetChecksumAsync(projectId, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_Checksum.cs:107]: we pass the desired top level project (or nothing) into the SolutionState, which computes the same way as before, just without the frozenSG stuff.

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs [hunk 7]
```diff
@@ -13,34 +13,24 @@
 
 namespace Microsoft.CodeAnalysis.Serialization;
 
-internal sealed class SolutionStateChecksums(
-    Checksum attributes,
-    ChecksumsAndIds<ProjectId> projects,
-    ChecksumCollection analyzerReferences,
+internal sealed class SolutionCompilationStateChecksums(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:16]: pulled compilation-state outwards.

File: src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Test.Next/Remote/SerializationValidator.cs:221]: but, basically, the tests are updated now that we both make a request for a the solution-comp-state, and then the solution-state.

**Comment by sharwell** [comment]
@CyrusNajmabadi  Does this change apply to VS Code / LSP?

**Comment by CyrusNajmabadi** [comment]
Yes. Checksums are also used in vscode/lsp for all features that want to be able to indicate "nothing changed" between requests.

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**